_posts/2020-01-09-rxjs-wstep.md

```diff
@@ -0,0 +1,180 @@
+
+Wywołanie funkcji _subscribe()_ dopisuje nas do listy obserwatorów danego strumienia.
+
+W momencie podpięcia się jako obserwator do danego Observable zostaje przekazany nam obiekt typu Subscription, na którym możemy wywołać metodę _unsubscribe()_, która usunie nas z listy obserwatorów. 
+
+Odsubskrybowanie się jest bardzo ważne w przypadku strumieni gorących, które w większości przypadków są nieskończone - emitują wartość przez potencjalnie nieskończoną ilość czasu. Jeżeli zapomnimy o usunięciu nas z listy obserwatorów danego strumienia, referencja do stworzonego przez nas obserwatora będzie istnieć przez cały cykl życia aplikacji - tworząc nieskończoną dziurę w pamięci, która w sytuacji ekstremalnej może doprowadzić do zabicia karty, w której działa nasza aplikacja.
+## Subject - tworzenie własnych strumieni
+RxJS oferuje nam również możliwość tworzenia własnych strumieni. Możemy to zrobić za pomocą obiektów typu Subject. Subject, tłumacząc na język polski, oznacza dosłownie: _temat_. Strumień taki jest więc tematem, potencjalnie nieskończonym, który emituje nowe wartości w kluczowych dla nas miejscach aplikacji. 
+Subject możemy stworzyć bardzo prosto - jak każdy inny obiekt:
+```typescript
+const subject$ = new Subject<number>();
+```
+Do danego strumienia możemy teraz się subskrybować jako obserwatorzy:
+```typescript
+subject$.asObservable()
+    .subscribe((value) => console.log('value from subject$: ', value))
+```
+Zmienna _subject$_ oferuje nam teraz metodę _next(value: number)_, dzięki której możemy rozesłać wszystkim obserwatorom nową wartość liczbową, w taki sposób:
+```typescript
+subject$.next(5);
+// value from subject$: 5
+``` 
+Strumień, jaki teraz stworzyliśmy jest strumieniem nieskończonym, więc musimy pamiętać o odsubskrybowaniu się po zakończeniu nasłuchiwania!
+Wyemitowaliśmy w powyższym przykładzie wartość numeryczną '5', którą otrzymał pojedynczy obserwator. Jeżeli w momencie emisji nowej wartości nie istniałby żaden obserwator, to wartość ta przepadła by na wieki. Jest to dosyć smutne. Co w przypadku, gdy ta wartość jest dla nas ważna i nie chcemy jej stracić? Na szczęście z pomocą przychodzą nam specyficzne obiekty rozszerzające typ Subject.
+### ReplaySubject
+ReplaySubject jest strumieniem, który dla każdego nowego obserwatora odtwarza N ostatnio emitowanych danych. Wartość N możemy przekazać w konstruktorze tego obiektu, o tak:
+```typescript
+const replaySubject$ = new ReplaySubject<number>(5);
+``` 
+Przy subskrypcji, jeżeli wcześniej emitowane były wartości przez ten strumień zostaną one odtworzone danemu obserwatorowi, ale nie więcej niż 5 ostatnich.
```

Comment:
Tutaj trzeba pozamieniać trochę wyrazy:
Przy subskrypcji, jeżeli wcześniej emitowane były wartości przez ten strumień zostaną one odtworzone danemu obserwatorowi, ale nie więcej niż 5 ostatnich. 
==>
Jeżeli wcześniej przez ten strumień emitowane były wartości, to przy subskrypcji, zostaną one odtworzone danemu obserwatorowi, ale nie więcej niż 5 ostatnich.
LUB
W przypadku subskrypcji, jeżeli wcześniej przez ten strumień emitowane były wartości, to zostaną one odtworzone danemu obserwatorowi, ale nie więcej niż 5 ostatnich.